features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/source/src/app/components/Apis/Create/GraphQL/APICreateGrpahQL.jsx

```diff
@@ -0,0 +1,369 @@
+            if (name === 'endpoint') {
+                changes.endpointConfig = {
+                    endpoint_type: 'http',
+                    sandbox_endpoints: {
+                        url: value,
+                    },
+                    production_endpoints: {
+                        url: value,
+                    },
+                };
+            } else {
+                changes[name] = value;
+            }
+
+
+            // Checking validity.
+
+            const validUpdated = valid;
+            validUpdated.name.empty = !api.name;
+            validUpdated.context.empty = !api.context;
+            validUpdated.version.empty = !api.version;
+            validUpdated.endpointConfig.empty = !api.endpointConfig;
+            // TODO we need to add the already existing error for (context)
+            // by doing an api call ( the swagger definition does not contain such api call)
+
+            return { api: changes, valid: validUpdated };
+        });
+    }
+
+    /**
+     * Make API POST call and create send GraphQL file or URL
+     * @memberof APICreateGraphQL
+     */
+    createGraphQLAPI() {
+        this.setState({ loading: true });
+        const newApi = new API();
+        const { graphQLBean, api } = this.state;
+        const {
+            name, version, context, policies, endpointConfig, gatewayEnvironments, implementationType, operations,
+        } = api;
+        const uploadMethod = 'file';
+        const apiAttributes = {
+            name,
+            version,
+            context,
+            endpointConfig,
+            gatewayEnvironments,
+            policies,
+            operations,
+        };
+        const apiData = {
+            additionalProperties: JSON.stringify(apiAttributes),
+            implementationType,
+            [uploadMethod]: graphQLBean[uploadMethod],
+            file: graphQLBean.file,
+        };
+
+        newApi
+            .importGraphQL(apiData)
+            .then((response) => {
+                Alert.success(`${name} API Created Successfully.`);
+                const uuid = response.obj.id;
+                const redirectURL = '/apis/' + uuid + '/overview';
+                this.setState({ loading: false });
+                this.props.history.push(redirectURL);
+            })
+            .catch((errorResponse) => {
+                this.setState({ loading: false });
+                console.error(errorResponse);
+                const error = errorResponse.response.obj;
+                const messageTxt = 'Error[' + error.code + ']: ' + error.description + ' | ' + error.message + '.';
+                Alert.error(messageTxt);
+            });
+    }
+    updateFileErrors(newValid) {
+        this.setState({ valid: newValid });
+    }
+
+    handleNext = () => {
+        const { activeStep, graphQLBean, valid } = this.state;
+        const { uploadMethod } = 'file';
+
+        const validNew = JSON.parse(JSON.stringify(valid));
```

Comment:
I using this deep copy is performance intensive, better to use [1] @tmkasun 


[1] - [https://lodash.com/docs/4.17.15#cloneDeep](https://lodash.com/docs/4.17.15#cloneDeep)

Reply:
Yes, it a problem, Best thing to avoid such coping is to keep only what you use/need in the local state or sometimes called state normalization, So you don't need to handle unwanted complex objects. Let's keep this in this way and moving forward, Try to avoid such JSON parse style copings